dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java

```diff
@@ -133,6 +141,122 @@
     @Getter
     private Map<String, PeriodType> periodTypeMap = new HashMap<>();
 
+    /**
+     * Internal map of category combo + category options (key) to category
+     * option combo (value).
+     *
+     * Category option combo value will be in the idScheme defined by the user
+     * on import.
+     */
+    private final Map<Pair<String, Set<String>>, String> cosToCOC = new HashMap<>();
+
+    /**
+     * Store mapping of category combo + category options identifiers(key) to
+     * category option combo identifiers (value).
+     *
+     * Category options, category option combo identifiers will be in the
+     * idScheme defined by the user on import. Note: different idSchemes for
+     * category combos are not supported.
+     */
+    public void putCategoryOptionCombo( CategoryCombo categoryCombo, Set<CategoryOption> categoryOptions,
+        CategoryOptionCombo categoryOptionCombo )
+    {
+        if ( categoryOptionCombo != null )
+        {
+            TrackerIdentifier optionComboIdScheme = getIdentifiers().getCategoryOptionComboIdScheme();
+            this.cosToCOC.put( categoryOptionComboCacheKey( categoryCombo, categoryOptions ),
+                optionComboIdScheme.getIdentifier( categoryOptionCombo ) );
+            this.put( optionComboIdScheme, categoryOptionCombo );
+        }
+        else
+        {
+            this.cosToCOC.put( categoryOptionComboCacheKey( categoryCombo, categoryOptions ), null );
+        }
+    }
+
+    private Pair<String, Set<String>> categoryOptionComboCacheKey( CategoryCombo categoryCombo,
+        Set<CategoryOption> categoryOptions )
+    {
+        Set<String> coIds = categoryOptions.stream()
+            .map( getIdentifiers().getCategoryOptionIdScheme()::getIdentifier )
+            .collect( Collectors.toSet() );
+        return Pair.of( categoryCombo.getUid(), coIds );
+    }
+
+    /**
+     * Check if a category option combo for given category combo and category
+     * options has been stored using {@link #putCategoryOptionCombo}. Returns
+     * true if null and a non-null category option combo have been stored.
+     *
+     * @param categoryCombo category combo
+     * @param categoryOptions category options
+     * @return true if category option combo has been stored given both
+     *         arguments
+     */
+    public boolean containsCategoryOptionCombo( CategoryCombo categoryCombo, Set<CategoryOption> categoryOptions )
+    {
+        return this.cosToCOC.containsKey( categoryOptionComboCacheKey( categoryCombo, categoryOptions ) );
+    }
+
+    /**
+     * Get the category option combo for given category combo and category
+     * options. For the category option combo to exist it has to be stored
+     * before using {@link #putCategoryOptionCombo}.
+     *
+     * @param categoryCombo category combo
+     * @param categoryOptions semi-colon separated list of category options
```

Comment:
It is a semi-colon separated list since that is what we expect in the payload: [event.attributeCategoryOptions](https://github.com/dhis2/dhis2-core/blob/3a6cf1a28322c9c34ac6f8fdaf210615fa4e3567/dhis-2/dhis-services/dhis-service-tracker/src/test/resources/tracker/event_events.json#L12).

Reply:
Good point, I will fix that. The payload contains a string field with semi-colon separated identifiers for category options. So this sentence is correct on the other getter where you can pass in the user input (string). Here its just a set of CategoryOption 😄